documentation/src/main/asciidoc/reference/introduction.adoc

```diff
@@ -977,26 +977,54 @@ sessionFactory.withTransaction( (session, tx) -> session.persist(book) )
 
 This is probably the most convenient thing to use most of the time.
 
-=== Vert.x context
+== Integrating with Vert.x
 
 :vertx-context-introduction: https://vertx.io/blog/an-introduction-to-the-vert-x-context-object/
-
-The session is not thread-safe, using it in different threads may cause issues that are hard to debug.
-When it's created, a session is bound to a {vertx-context-introduction}[Vert.x context]
-and needs to be used in the same context after that.
-
-Hibernate Reactive will take care of this when the application uses `withSession`,
-`withStatelessSession` or `withTransaction` (See <<_obtaining_a_reactive_session>>).
-
-When using `openSession` or `openStatelessSession`, one need to make sure
-to use it in the right Vert.x context or the following exception might occur:
+:vertx-local-context: https://vertx.io/docs/apidocs/io/vertx/core/Context.html#getLocal-java.lang.Object-
+:vertx-get-context: https://vertx.io/docs/apidocs/io/vertx/core/Vertx.html#getOrCreateContext--
+:vertx-runon-context: https://vertx.io/docs/apidocs/io/vertx/core/Context.html#runOnContext-io.vertx.core.Handler-
+
+At runtime, interaction with the database happens on a Vert.x thread,
+typically the event loop thread. When you write code that creates and
+destroys Hibernate Reactive sessions, it's important to understand how
+sessions relate to threads and {vertx-context-introduction}[Vert.x contexts].
+
+=== Sessions and Vert.x contexts
+
+TIP: Remember how in regular old Hibernate JPA, you're not supposed to
+share a session between multiple threads? Well, the idea here is essentially
+similar, it's just that the notion of a "thread" is a little more slippery,
+or at least more _technical_. You need to be able to replace the idea of a
+"thread" with the idea of a chain of callbacks occurring on a reactive stream,
+all within the scope of a certain Vert.x _local context_.
+
+When you create a session using `withSession()` or `withTransaction()`, it's
+automatically associated with the current Vert.x
+{vertx-local-context}[local context], and propagates with the local context,
+as mentioned above in <<_obtaining_a_reactive_session>>. And you're only
+allowed to use the session from the thread that owns this local context. If
+you screw up, and use it from a different thread, you might see this error:
 
 [source, JAVA, indent=0]
 ----
 HR000068: This method should exclusively be invoked from a Vert.x EventLoop thread; ...
 ----
 
-One way to run the code in the correct Vert.x context is the following:
+On the other hand, if you use `openSession()`, you'll have to manage the
+association between sessions and contexts yourself. Now, that's in principle
+straightforward, but you'd be surprised how often people mess up.
+
+IMPORTANT: The session is not thread-safe (or "stream-safe"), so using it across
+different threads (or reactive streams) may cause bugs that are _extremely_ hard
+to detect. Don't say we didn't warn you!
```

Comment:
I think we should add a quick example of what it means to use the session on different reactive streams.
For example a loop that persist multiple entities:
```
List<CompletionStage> list = ...
for (Entity entity : entities) {
    list.add(session.persist(entity));
}
CompletableFuture.allOf( list ).thenCompose( session::flush );
```

Otherwise it's hard to wrap the head around when the warning applies

Reply:
Yup, agreed. Though I'm not sure I completely understand the example you've given myself :-)

Reply:
Shouldn't it be:
```
CompletionStage<Void> stage = ...
for (Entity entity : entities) {
    stage = stage.thenCompose( v -> session.persist(entity));
}
stage.thenCompose( session::flush );
```
?

Reply:
I guess I never looked at `CompletableFuture.allOf()` :-)

Reply:
I've used `allOf` because it seems like something useful but it  isn't this wrong already?
```
    list.add(session.persist(entity));
```
The `persist` are not chained and they will start as soon as the CompletionStage is created.

Reply:
Right, OK, then it's a good example.

But perhaps we should have one for Mutiny?

Reply:
> but isn't this already wrong?

Yeah well that's actually what confused me about the example: there's kinda two errors in it, so I wasn't clear that it was demonstrating the particular error we were discussing in the text.

Reply:
OK

> But perhaps we should have one for Mutiny?

Sure, it was just the first thing I thought

Reply:
Just put a bit of thought into this, and frankly I'm not sure I understand how these APIs do concurrency well enough to be able to write a good example (i.e. one that isn't wrong in *multiple* ways). So I'm going to leave the example out for now, until I think I know WTF I'm talking about.

Reply:
Can't we just use the example I mentioned, explain the two issues with it and then replace it when we find something better?